media-proxy/src/mesh/st2110.cc

```diff
@@ -0,0 +1,153 @@
+    case MESH_AUDIO_SAMPLE_RATE_96000:
+        return ST30_SAMPLING_96K;
+    case MESH_AUDIO_SAMPLE_RATE_44100:
+        return ST31_SAMPLING_44K;
+    default:
+        return ST30_SAMPLING_MAX;
+    }
+}
+
+st30_ptime ST2110::mesh_audio_ptime_to_st_ptime(int ptime)
+{
+    switch (ptime) {
+    case MESH_AUDIO_PACKET_TIME_1MS:
+        return ST30_PTIME_1MS;
+    case MESH_AUDIO_PACKET_TIME_125US:
+        return ST30_PTIME_125US;
+    case MESH_AUDIO_PACKET_TIME_250US:
+        return ST30_PTIME_250US;
+    case MESH_AUDIO_PACKET_TIME_333US:
+        return ST30_PTIME_333US;
+    case MESH_AUDIO_PACKET_TIME_4MS:
+        return ST30_PTIME_4MS;
+    case MESH_AUDIO_PACKET_TIME_80US:
+        return ST31_PTIME_80US;
+    case MESH_AUDIO_PACKET_TIME_1_09MS:
+        return ST31_PTIME_1_09MS;
+    case MESH_AUDIO_PACKET_TIME_0_14MS:
+        return ST31_PTIME_0_14MS;
+    case MESH_AUDIO_PACKET_TIME_0_09MS:
+        return ST31_PTIME_0_09MS;
+    default:
+        return ST30_PTIME_MAX;
+    }
+}
+
+void *ST2110::get_frame_data_ptr(st_frame *src) { return src->addr[0]; }
+void *ST2110::get_frame_data_ptr(st30_frame *src) { return src->addr; }
+
+int ST2110::frame_available_cb(void *ptr)
+{
+    auto _this = static_cast<ST2110 *>(ptr);
+    if (!_this) {
+        return -1;
+    }
+
+    _this->_stop.store(true);
+    _this->_cv.notify_all();
+
+    return 0;
+}
+
+void ST2110::get_mtl_dev_params(mtl_init_params &st_param, const std::string &dev_port,
+                                mtl_log_level log_level,
+                                const char local_ip_addr[MESH_IP_ADDRESS_SIZE])
+{
+    if (getenv("KAHAWAI_CFG_PATH") == NULL) {
+        setenv("KAHAWAI_CFG_PATH", "/usr/local/etc/imtl.json", 0);
+    }
+    strlcpy(st_param.port[MTL_PORT_P], dev_port.c_str(), MTL_PORT_MAX_LEN);
+    inet_pton(AF_INET, local_ip_addr, st_param.sip_addr[MTL_PORT_P]);
+    st_param.pmd[MTL_PORT_P] = mtl_pmd_by_port_name(st_param.port[MTL_PORT_P]);
+    st_param.num_ports = 1;
+    st_param.flags = MTL_FLAG_BIND_NUMA;
+    st_param.flags |= MTL_FLAG_TX_VIDEO_MIGRATE;
+    st_param.flags |= MTL_FLAG_RX_VIDEO_MIGRATE;
+    st_param.flags |= MTL_FLAG_RX_UDP_PORT_ONLY;
+    st_param.pacing = ST21_TX_PACING_WAY_AUTO;
+    st_param.log_level = log_level;
+    st_param.priv = NULL;
+    st_param.ptp_get_time_fn = NULL;
+    // Native af_xdp have only 62 queues available
+    if (st_param.pmd[MTL_PORT_P] == MTL_PMD_NATIVE_AF_XDP) {
+        st_param.rx_queues_cnt[MTL_PORT_P] = 62;
+        st_param.tx_queues_cnt[MTL_PORT_P] = 62;
+    } else {
+        st_param.rx_queues_cnt[MTL_PORT_P] = 128;
+        st_param.tx_queues_cnt[MTL_PORT_P] = 128;
+    }
+    st_param.lcores = NULL;
+    st_param.memzone_max = 9000;
+}
+
+mtl_handle ST2110::get_mtl_handle(const std::string &dev_port, mtl_log_level log_level,
+                                  const char local_ip_addr[MESH_IP_ADDRESS_SIZE])
+{
+    static mtl_handle dev_handle;
+    static std::mutex mtx;
```

Comment:
What is a shared resource protected by this mutex? Is it the same resource in ::configure()?

Reply:
Shared resource is mtl_handle, it can be initialized only once for whole application/process